docs/fsharp/language-reference/type-extensions.md

```diff
@@ -1,99 +1,151 @@
+It is not possible to define an optional extension on a generic type when the type variable is constrained. For example, the following is not possible:
 
 ```fsharp
 open System.Collections.Generic
 
+// NOT POSSIBLE!
+// Seq.sum constraints 'T.
 type IEnumerable<'T> with
-    /// Repeat each element of the sequence n times
-    member xs.RepeatElements(n: int) =
-        seq { for x in xs do for i in 1 .. n do yield x }
+    member this.Sum() = Seq.sum this
 ```
 
-However, for a generic type, the type variable may not be constrained. You can now declare a C#-style extension member in F# to work around this limitation. When you combine this kind of declaration with the inline feature of F#, you can present generic algorithms as extension members.
+To get around this restriction, you can use extension methods.
+
+## Extension methods
 
-Consider the following declaration:
+Finally, *extension methods* (sometimes called "C# style extension members") can be declared in F# as either a let-bound value value in a module or a member on a class.
+
+Extension methods are useful for when you wish to define extensions on a generic type that will constrain the type variable. For example:
 
 ```fsharp
+namespace Extensions
+
+open System.Runtime.CompilerServices
+
 [<Extension>]
-type ExtraCSharpStyleExtensionMethodsInFSharp () =
+module IEnumerableExtensions =
     [<Extension>]
-    static member inline Sum(xs: IEnumerable<'T>) = Seq.sum xs
+    let inline Sum(xs: IEnumerable<'T>) = Seq.sum xs
 ```
 
-By using this declaration, you can write code that resembles the following sample.
+When used, this code will make it appear as if `Sum` is defined on <xref:System.Collections.Generic.IEnumerable`1>, so long as `Extensions` has been opened or is in scope.
 
-```fsharp
-let listOfIntegers = [ 1 .. 100 ]
-let listOfBigIntegers = [ 1I to 100I ]
-let sum1 = listOfIntegers.Sum()
-let sum2 = listOfBigIntegers.Sum()
-```
+## Other remarks
+
+Type extensions also have the following attributes:
 
-In this code, the same generic arithmetic code is applied to lists of two types without overloading, by defining a single extension member.
+* Any type that can be accessed can be extended.
+* Intrinsic and optional type extensions can define _any_ member type, not just methods. So extension properties are also possible, for example.
+* The `self-identifier` token in the [syntax](type-extensions.md#syntax) represents the instance of the type being invoked, just like ordinary members.
+* Extended members can be static or instance members.
 
+The following limitations also exist for type extensions:
+
+* Type extensions cannot be virtual or abstract methods.
+* Type extensions cannot be defined on [type abbreviations](type-abbreviations.md).
+* You can define extensions that overload other methods of the same name, but the F# compiler gives preference to non-extension methods in the case of an ambiguous call.
```

Comment:
Byref 'this' is not possible due to self identifier rigidity, no attributes possible etc

Reply:
Yep, `byref` is not possible (I've tried, unsuccessfully) - will call that out

Reply:
Or rather, you can definitely declare one like this:

```fsharp
type byref<'T> with
    member __.Boof() = ()
```

But usage doesn't seem possible:

<img width="510" alt="screen shot 2018-07-14 at 21 13 12" src="https://user-images.githubusercontent.com/6309070/42730516-cfdb1bcc-87aa-11e8-9c82-4a7e140dd2e3.png">

And fundamentally, the concept of extending a byref simply doesn't make sense. Adding arbitrary members to a pointer should simply not be possible. And in the extensive byref work we did to properly support span, I believe we are disallowing this concept altogether.

Reply:
Hmm, extending attributes seems fine as a declaration, though I wonder how you'd make any use of this:

<img width="244" alt="screen shot 2018-07-14 at 21 46 41" src="https://user-images.githubusercontent.com/6309070/42730683-69af2e1a-87af-11e8-9706-a5979cadcfbe.png">

Reply:
I meant a by reference 'this', such a feature is very useful for struct (and ref struct) extensions to skip a copy on invocation. VB always had it and C# with all the ref return/local perf features is considering following suit https://github.com/dotnet/csharplang/issues/186